### DISCUSSION TITLE
First day with BudiBase

### DISCUSSION THREAD

**Discussion by Pachat** [Budibase/budibase#2683] [Help]
Window installation was a comfortable experience, well documented, Thanks.

But I have still some questions, maybe because I'm discovering BudiBase as of today:

- The popup (Create, Modifiy, etc.) should all be modals. This would also allow not to loose all what is entered in case of error.
- When there is a text length limit, the user should not be allowed to type more chars
- Cannot hide columns from table and show only in forms.
- Did not know what to do after clicking on Publish. Then found the > button. May be the order Publish > (instead of > Publish) is more intuitive.
- Click on a relation field in the table does not open the related row form
- What is the effect of "Use as table display column"?
- How do I hide a column in a view table and not in another?
- When I delete a column in a view, it deletes the column and the data: this should not be the case. A view shows data but should not act on the data structure.
- Date format should be customizable. Date could be entered by typing. The calendar is tedious for far away dates.
- Cannot change tabs' labels (even if table name is changed)
- Cannot reorder tabs (best would be with drag and drop)
- Can I place labels above fields
- Can I have direct access to the CouchDB? 
- I plan to use MariaDB : is it possible? Will it be a total replacement of CouchDB storing even the App?
- Is there a customizable template file?
- Is there a kind of user.js that allows to introduce more actions, triggers, etc.?
- Where are the .css located? Is there a user.css that is loaded at last?
- Can I use VCS (git)? (Is all the App in database or also in some files)

**Comment by mjashanks** [comment]
Hi @Pachat. Thanks for taking the time to communicate this - very much appreciated!

> The popup (Create, Modifiy, etc.) should all be modals. This would also allow not to loose all what is entered in case of error.

You can show any screen inside a modal - this explains how: https://docs.budibase.com/guides/crud-app-with-budibase-db/creating-a-new-service-log-with-a-modal

> When there is a text length limit, the user should not be allowed to type more chars

Agreed! This is an issue.

> Cannot hide columns from table and show only in forms.

You can do this - the table component has a "Columns" setting:
![image](https://user-images.githubusercontent.com/3524181/134138344-ece1bf86-e67c-4699-8b9a-a0fe5ff41cae.png)

> Did not know what to do after clicking on Publish. Then found the > button. May be the order Publish > (instead of > Publish) is more intuitive.

So, the `>` previews your app - notice the URL in this instance is `/app_dev_`. However, if anyone is currently using your app, they will not see the changes that you have just made, until you click `Publish`. 

When you click publish, you can go back to your App List screen, click the `...`beside your app and "View published app". You will notice that the URL for this is `/app_`. This is the URL that your users log into.

![image](https://user-images.githubusercontent.com/3524181/134138895-4fb2e454-9ab9-4157-8a7f-2a4abee08d7b.png)

Note that if you create a user who does not have "Development access", they see a different Apps List. When they click on an app, it brings them directly to the Published App. You can see this mode using the "Close developer mode" option, in the app list screen:

![image](https://user-images.githubusercontent.com/3524181/134139119-9c6aed41-2c77-42ec-a738-78a5ddf1ab4f.png)


> Click on a relation field in the table does not open the related row form

Thanks - we know about this, and it's logged!

> What is the effect of "Use as table display column"?

When you create a relationship, this Is how that row displays in the related table. In my screenshot below, I have a "Contacts" column - this is a relationship to another table. The values of that column ("Mike") come from the "Display Column" of the Contacts table.
![image](https://user-images.githubusercontent.com/3524181/134139864-f7e29e79-00d6-4b8e-9f4d-4d0c4980e232.png)

> How do I hide a column in a view table and not in another?

What do you mean by a "view table"? As above, when you add a table component to your frontend, you can choose which columns to display.

> When I delete a column in a view, it deletes the column and the data: this should not be the case. A view shows data but should not act on the data structure.

We should not allow columns to be deleted in the view at all.

Note that you are referring to the user interface within the Budibase builder, which is NOT the user interface that your users will see, when using the app that you build.

> Date format should be customizable. 

you can customise the date formatting in the design section when you display the date inside text components. However, I believe that this is not possible inside the date picker, which I think you might be referring to?

> Date could be entered by typing. 

Good point!
> The calendar is tedious for far away dates.

Good point!

> Cannot change tabs' labels (even if table name is changed)

Im not sure what you mean - what are the "tabs' labels" ?

> Cannot reorder tabs (best would be with drag and drop)

Again - what are tabs?

> Can I place labels above fields?

Yes. You can remove the label from the field, and insert a "Paragraph" component above the field (your label). 

> Can I have direct access to the CouchDB?

yes - https://github.com/Budibase/budibase/discussions/2009 

> I plan to use MariaDB : is it possible? Will it be a total replacement of CouchDB storing even the App?

Yes - you can use Maria DB for your data, by adding a MySQL datasource:
![image](https://user-images.githubusercontent.com/3524181/134141190-dbe6220e-db29-4649-b56b-d7e624cdab3f.png)

You can't totally replace CouchDB - all the Budibase Metadata is stored here - e.g. Apps, Users, Config etc

> Is there a customizable template file?

Can you clarify what you mean by "customizable template file"? What do you wish to template? 

> Is there a kind of user.js that allows to introduce more actions, triggers, etc.?

Not yet - we hope to add this soon - can't give you a date though.

> Where are the .css located? Is there a user.css that is loaded at last?

The web files are all stored inside MiniIO, which is running on one of the docker containers. Theres no supported way of modifying the CSS files. In our upcoming release, you will be able to declare some basic custom themes. Also, on each component there is an "Edit Custom CSS" button, which gives you some flexibility over individual components.

> Can I use VCS (git)? (Is all the App in database or also in some files) 

No - everything is stored in CouchDB - we will be adding revision history at some point, but it wont be VCS as your used to.

**Comment by Pachat** [comment]
Hi @mjashanks,
I really appreciate your detailed and precise answer. It makes me confident that the product is well supported.
I've been able to go further today.

Facing some issues,

- If I change a column name in my external database, and click in the admin panel to show the table it hangs: to avoid this, it is necessary to again Fetch tables from database first.
- Also, i faced  #2377  but found that it is a "good first issues" and will be resolved soon, like
- I read somewhere that adding javascript will come end of October. That's great.
- By "tab", I meant: what to do to change the order Contacts / Events to Events / Contacts,:
![image](https://user-images.githubusercontent.com/13746133/134244051-d3090bd0-5b08-44d8-81fd-35d15c120ef8.png)

**Reply by mjashanks** [reply]
Thanks @Pachat .

You can change the tabs under "Layouts" - You set the "Layout" of a screen, which is like a "Master Page". This is where you would put some common headers and footers - i.e. the navigation tabs.

You can find the Layouts here:
![image](https://user-images.githubusercontent.com/3524181/134309254-ff1c6a6e-a094-4efd-b381-42e660f1a76d.png)

So, select which layouts you which to change (Navigation Layout), and click "Configure Links" from the right-hand side.
